server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.druid.server.coordinator.stats.RowKey;
+import org.apache.druid.server.coordinator.stats.Stats;
+import org.apache.druid.timeline.DataSegment;
+import org.apache.druid.timeline.SegmentId;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manages information about replication status of segments in a cluster.
+ */
+@LazySingleton
+public class SegmentReplicationStatusManager
+{
+  private final MetadataManager metadataManager;
+
+  /**
+   * Set of broadcast segments determined in the latest coordinator run of the {@link RunRules} duty.
+   * This might contain stale information if the Coordinator duties haven't run or are delayed.
+   */
+  private volatile Set<DataSegment> broadcastSegments = null;
+
+  /**
+   * Used to determine count of under-replicated or unavailable segments.
+   * Updated in each coordinator run in the {@link SegmentReplicationStatusManager.UpdateReplicationStatus} duty.
+   * <p>
+   * This might have stale information if coordinator runs are delayed. But as
+   * long as the {@link SegmentsMetadataManager} has the latest information of
+   * used segments, we would only have false negatives and not false positives.
+   * In other words, we might report some segments as under-replicated or
+   * unavailable even if they are fully replicated. But if a segment is reported
+   * as fully replicated, it is guaranteed to be so.
+   */
+  private volatile SegmentReplicationStatus segmentReplicationStatus = null;
+
+  @Inject
+  public SegmentReplicationStatusManager(MetadataManager metadataManager)
+  {
+    this.metadataManager = metadataManager;
+  }
+
+  /**
+   * @return Set of broadcast segments determined by the latest run of the {@link RunRules} duty.
+   * If the coordinator runs haven't triggered or are delayed, this information may be stale.
+   */
+  @Nullable
+  public Set<DataSegment> getBroadcastSegments()
+  {
+    return broadcastSegments;
+  }
+
+  public Object2IntMap<String> getDatasourceToUnavailableSegmentCount()
+  {
+    if (segmentReplicationStatus == null) {
+      return Object2IntMaps.emptyMap();
+    }
+
+    final Object2IntOpenHashMap<String> datasourceToUnavailableSegments = new Object2IntOpenHashMap<>();
+
+    final Iterable<DataSegment> dataSegments = metadataManager.segments().iterateAllUsedSegments();
+    for (DataSegment segment : dataSegments) {
+      SegmentReplicaCount replicaCount = segmentReplicationStatus.getReplicaCountsInCluster(segment.getId());
+      if (replicaCount != null && (replicaCount.totalLoaded() > 0 || replicaCount.required() == 0)) {
+        datasourceToUnavailableSegments.addTo(segment.getDataSource(), 0);
+      } else {
+        datasourceToUnavailableSegments.addTo(segment.getDataSource(), 1);
+      }
+    }
+
+    return datasourceToUnavailableSegments;
+  }
+
+  public Object2IntMap<String> getDatasourceToDeepStorageQueryOnlySegmentCount()
```

Comment:
@findingrish , this seems like the only new method that has been added here.
Please remove this new class `SegmentReplicationStatusManager` and move the code back to`DruidCoordinator`.

If a refactor is required, please do it in a separate PR. 
This PR should focus only on the required changes.

Reply:
Correction: It seems that this method had already existed too. 
@findingrish , is there any new code in `SegmentReplicationStatusManager`?

Reply:
@kfaraz There is no new code in `SegmentReplicationStatusManager`. The reason for refactoring was a cyclic dependency between `CoordinatorSegmentMetadataCache` and `DruidCoordinator` while trying to use `DruidCoordinator#getSegmentReplicationFactor`. 

I will raise a separate PR for the refactor.

Reply:
Okay. Can you share some more details on how the cyclic dependency is coming into picture?

Reply:
Currently `DruidCoordinator` has a dependency on `CoordinatorSegmentMetadataCache`, for this patch I need to use `DruidCoordinator#getSegmentReplicationFactor` in `CoordinatorSegmentMetadataCache` which is resulting in cyclic dependency. 

As a solution, I have refactored `DruidCoordinator` to separate out the code which updates `segmentReplicationStatus` and `broadcastSegments`.

Reply:
@findingrish , you could just expose a method `updateSegmentReplicationStatus()` on `CoordinatorSegmentMetadataCache`. Call this method from `DruidCoordinator.UpdateReplicationStatus.run()` where we update `broadcastSegments` and `segmentReplicationStatus`.

Let me know if this works for you.

Reply:
Yeah, this approach would work for me. 
However, it seems bit odd that `DruidCoordinator.UpdateReplicationStatus` has to additionally update state in some other class, ideally the consumer `CoordinatorSegmentMetadataCache` should be pulling this information? 

Is there a reason to avoid the refactor work?

Reply:
> Is there a reason to avoid the refactor work?

Yes, the dependencies are already all over the place which makes the code less readable and also complicates testing. A refactor is needed here but it would have to be thought through a little.

> However, it seems bit odd that DruidCoordinator.UpdateReplicationStatus has to additionally update state in some other class,

Not really, you can think of the `DruidCoordinator` (or rather the `UpdateReplicationStatus` duty in this case) as sending a notification to the `CoordinatorSegmentMetadatCache` that the segment replication status has been updated. The `DruidCoordinator` already sends notification to the metadata cache about leadership status, this is another notification in the same vein.

Reply:
> Yes, the dependencies are already all over the place which makes the code less readable and also complicates testing. A refactor is needed here but it would have to be thought through a little.

Yes, this makes sense. `DruidCoordinator` refactoring would need more thought. 

Thanks for the suggestion, I will update the patch.